docs: reformatting tables in drivers/gm.adoc (smoe:docs_drivers_gm_tables_reformatting) (#2037)
LinuxCNC/linuxcnc: docs/src/drivers/gm.adoc

> gm.<nr of card>.gpio.<nr of gpio con>

Reviewer: The underscores are not really needed here. Maybe even misleading as it look then more like a defined name or parameter.

> gm.<nr of card>.gpio.<nr of gpio con>

Reviewer 2: Not needed per se, but as I said, it's current recommended style and also avoid suggesting spaces are allowed.
Plus, in this very case, this is somehow a defined value.

> gm.<nr of card>.gpio.<nr of gpio con>

Reviewer: Recommended by the Gitlab style guide?

> gm.<nr of card>.gpio.<nr of gpio con>

Reviewer 2: Not explicitly but the example suggests it https://docs.gitlab.com/ee/development/documentation/styleguide/#placeholder-text

Same here http://courses.cms.caltech.edu/cs11/material/general/usage.html

And the very nice https://redhat-documentation.github.io/asciidoc-markup-conventions/ (parameters and variables at the bottom)

> gm.<nr of card>.gpio.<nr of gpio con>

Reviewer: Okay, persuaded :slightly_smiling_face:

----

where _<nr of gpio con>_ is from 0 to 3.
For example, `gm.0.gpio.0.in-0` indicates the state of the

> first pin of the first GPIO connector on the GM6-PCI card.

Reviewer: Yeah but this is not a code example, just a pin name.

> first pin of the first GPIO connector on the GM6-PCI card.

Reviewer 2: hmm, not a bad point... But I have the feeling that for someone configuring a machine and having to dive into documentation, the examples are almost as important to identify quickly than the syntax pattern, all the more in the beginning when she needs it most.

m| .min-speed-estimate | (float, R/W)        |
Determine the minimum measured velocity magnitude at which .velocity will be set as nonzero.
Setting this parameter too low will cause it to take a long time for velocity to go to zero after encoder pulses have stopped arriving.
m| .position-scale     | (float, R/W)        |
Scale in counts per length unit. .position=.counts/.position-scale.
For example, if position-scale is 2000, then 1000 counts of the encoder will produce a position of 0.5 units.
|===

.HAL example

Setting encoder module of axis 0 to receive 500 CPR quadrature encoder signal and use reset to round position.

[options="header"]
|===
| setp pin setting                           | Comment
m|setp gm.0.encoder.0.counter-mode 0         | 0: quad, 1: stepDir
m|setp gm.0.encoder.0.index-mode 1           | 0: reset pos at index, 1:round pos at index
m|setp gm.0.encoder.0.counts-per-rev 2000    | GM process encoder in 4x mode, 4x500=2000
m|setp gm.0.encoder.0.index-invert 0         |
m|setp gm.0.encoder.0.min-speed-estimate 0.1 | in position unit/s
m|setp gm.0.encoder.0.position-scale 20000   | 10 encoder rev cause the machine to move one position unit (10x2000)

> |===

Author: done - thanks!

l| .index-enabled     | (bit, IO)          |
When True, counts and position are rounded or reset
(depends on index-mode) on next rising edge of channel-I.
Every time position is reset because of Index, index-enabled
pin is set to 0 and remain 0 until connected HAL pin does
not set it.
l| .velocity          | (float, Out)       |
Velocity in scaled units per second. GM encoder uses high
frequency hardware timer to measure time between encoder
pulses in order to calculate velocity. It greatly reduces
quantization noise as compared to simply differentiating
the position output. When the measured velocity is below
min-speed-estimate, the velocity output is 0.
|===

.Encoder parameters
[width="80%",options="header",cols="<3,^2,<6"]
|===
| Parameters          | Type and Read/Write | Parameter description
l| .counter-mode       | (bit, R/W)          |
When True, the counter counts each rising edge of the
channel-A input to the direction determined by channel-B.
This is useful for counting the output of a single channel
(non-quadrature) or step/dir signal sensor. When false, it
counts in quadrature mode.
l| .index-mode         | (bit, R/W)          |

> When True and .index-enabled is also true, .counts and
> .position are rounded (based on .counts-per-rev) at rising

Reviewer: ```suggestion
When True and _.index-enabled_ is also true, _.counts_ and
_.position_ are rounded (based on _.counts-per-rev_) at rising
```
Maybe even highlight them.

> .position are rounded (based on .counts-per-rev) at rising

Reviewer 2: These all should be put between backtick actually ! Missed that :/

l| .min-speed-estimate | (float, R/W)        |
Determine the minimum measured velocity magnitude at which
.velocity will be set as nonzero. Setting this parameter too
low will cause it to take a long time for velocity to go to zero
after encoder pulses have stopped arriving.
l| .position-scale     | (float, R/W)        |
Scale in counts per length unit. .position=.counts/.position-scale.
For example, if position-scale is 2000, then 1000 counts of the
encoder will produce a position of 0.5 units.
|===

.HAL example

Setting encoder module of axis 0 to receive 500 CPR quadrature encoder signal and use reset to round position.

[options="header"]
|===
| setp pin setting                           | Comment
l|setp gm.0.encoder.0.counter-mode 0         | 0: quad, 1: stepDir
l|setp gm.0.encoder.0.index-mode 1           | 0: reset pos at index, 1:round pos at index
l|setp gm.0.encoder.0.counts-per-rev 2000    | GM process encoder in 4x mode, 4x500=2000
l|setp gm.0.encoder.0.index-invert 0         |
l|setp gm.0.encoder.0.min-speed-estimate 0.1 | in position unit/s
l|setp gm.0.encoder.0.position-scale 20000   | 10 encoder rev cause the machine to move one position unit (10x2000)

> |===

Reviewer: I totally agree.

> Connect encoder position to LinuxCNC position feedback:
> 

Reviewer 2: ```suggestion
.Connect encoder position to LinuxCNC joint position feedback
```

  edge of channel-I signal, otherwise on rising edge.
l| .min-speed-estimate | (float, R/W)        |
Determine the minimum measured velocity magnitude at which
.velocity will be set as nonzero. Setting this parameter too
low will cause it to take a long time for velocity to go to zero
after encoder pulses have stopped arriving.
l| .position-scale     | (float, R/W)        |
Scale in counts per length unit. .position=.counts/.position-scale.
For example, if position-scale is 2000, then 1000 counts of the
encoder will produce a position of 0.5 units.
|===

.HAL example

Setting encoder module of axis 0 to receive 500 CPR quadrature encoder signal and use reset to round position.

[options="header"]
|===
| setp pin seetting                          | Comment
l|setp gm.0.encoder.0.counter-mode 0         | 0: quad, 1: stepDir
l|setp gm.0.encoder.0.index-mode 1           | 0: reset pos at index, 1:round pos at index
l|setp gm.0.encoder.0.counts-per-rev 2000    | GM process encoder in 4x mode, 4x500=2000
l|setp gm.0.encoder.0.index-invert 0         |
l|setp gm.0.encoder.0.min-speed-estimate 0.1 | in position unit/s

> l|setp gm.0.encoder.0.position-scale 20000   | 10 encoder rev cause the machine to

Reviewer: Formatting this as table looses the connection that it is a code example

Author: done

> gm.<nr of card>.stepgen.<nr of axis>

Reviewer 2: ```suggestion
gm.<card_no>.stepgen.<axis_no>
```

|===

.StepGen module parameters
[width="90%",options="header",cols="<3,^2,<6"]
|===
| Parameters         | Type and Read/Write | Parameter description

> l| .step-type         | (u32, R/W)          | When 0, module produces Step/Dir signal. When 1, it

Reviewer: ```suggestion
m| .step-type         | (u32, R/W)          | When 0, module produces Step/Dir signal. When 1, it
```
"l" produces this ugly space...

Author: I have addressed these all locally.

.HAL example

.Setting StepGen module of axis 0 to generate 1000 step pulse per position unit

> [source,{hal}]]

Reviewer: It's one closing bracket to much here
```suggestion
[source,{hal}]
```

Setting StepGen module of axis 0 to generate 1000 step pulse per position unit:

[options="header"]
|===
| setp pin seetting                  | Comment
l|setp gm.0.stepgen.0.step-type 0    | 0:stepDir, 1:UpDown, 2:Quad
l|setp gm.0.stepgen.0.control-type 0 | 0:Pos. control, 1:Vel. Control
l|setp gm.0.stepgen.0.invert-step1 0 |
l|setp gm.0.stepgen.0.invert-step2 0 |
l|setp gm.0.stepgen.0.maxvel 0       |do not set maxvel for step lenerator, let interpolator control it.
l|setp gm.0.stepgen.0.maxaccel 0     |do not set max acceleration for step generator, let interpolator control it.
l|setp gm.0.stepgen.0.position-scale 1000 |1000 step/position unit
l|setp gm.0.stepgen.0.steplen 1000   | 1000 ns = 1 µs
l|setp gm.0.stepgen.0.stepspace 1000 | 1000 ns = 1 µs
l|setp gm.0.stepgen.0.dirdelay 2000  | 2000 ns = 2 µs

> |===

Reviewer: Formatting this as table looses the connection that it is a code example. I would rather add `[source,{hal}]` to the code block.

Author: Hm. Only the left side is executable, not the comment on the right.

Reviewer: >Agree with @hansu so same comment and proposal as above.
"Right side" could be converted to # <n> callouts but not convinced it would be beneficial here.

👍

[width="80%",options="header",cols="<3,^2,<6"]
|===
| Pins                         | Type and direction | Pin description
| gm.<nr of card>.power-enable | (bit, In)          |
If this pin is True, +
* and Watch Dog Timer is not expired +
* and there is no power fault

then power enable pins of axis- and CAN connectors are set to high, otherwise set to low.

> | gm.<nr of card>.power-fault  | (bit, Out)         | Power fault input.

Reviewer 2: ```suggestion
| gm.<card_no>.power-fault  | (bit, Out)         | Power fault input.
```

> gm.<nr of card>.dac.<nr of axis>
> ----
> 
> where _<nr of axis>_ is from 0 to 5. For example, `gm.0.dac.0.value`
> refers to the output voltage of DAC module of axis 0.
> 

Reviewer 2: ```suggestion
gm.<card_no>.dac.<axis_no>
----

Where _<axis_no>_ is from 0 to 5. +
For example:

.First DAC module output voltage on card 0
[source,{hal}]
----
gm.0.dac.0.value
----

```

> gm.<nr of card>.can-gm.<nr of axis>
> ----
> 
> where _<nr of axis>_ is from 0 to 5.
> For example, `gm.0.can-gm.0.position` refers to the output position of axis 0 in position units.

Reviewer 2: ```suggestion
gm.<card_no>.can-gm.<axis_no>
----

Where _<axis_no>_ is from 0 to 5. +
For example:

.Output position in position units on Axis 0 on card 0
[source,{hal}]
----
gm.0.can-gm.0.position
----
```

=== Parameters

.Watchdog parameters
[width="100%",options="header",cols="<3,^2,<6"]
|===
| Parameters                            | Type and direction | Parameter description

> | gm._<nr of card>_.watchdog-enable     | (bit, R/W)         |

Reviewer 2: ```suggestion
| gm._<card_no>_.watchdog-enable     | (bit, R/W)         |
```

Enable watchdog timer.
It is strongly recommended to enable watchdog timer,
because it can disables all the servo amplifiers by pulling down all enable signal in case of PC error.

> | gm._<nr of card>_.watchdog-timeout-ns | (float, R/W)       |

Reviewer 2: ```suggestion
| gm._<card_no>_.watchdog-timeout-ns | (float, R/W)       |
```

> Time interval in within the gm._<nr of card>_.read function must be executed.
> The gm._<nr of card>_.read is typically added to servo-thread,
> so watch timeout is typically set to 3 times of the servo period.

Reviewer 2: ```suggestion
Time interval in within the gm._<card_no>_.read function must be executed.
The gm._<card_no>_.read is typically added to servo-thread,
so watch timeout is typically set to 3 times of the servo period.
```

> gm.<nr of card>.joint.<nr of axis>
> ----
> 
> where _<nr of axis>_ is from 0 to 5. For example,
> `gm.0.joint.0.home-sw-in`
> indicates the state of the axis 0 home switch.

Reviewer 2: ```suggestion
gm.<card_no>.joint.<axis_no>
----

Where _<axis_no>_ is from 0 to 5. +
For example:

.Home switch state on Axis 0 on card 0
[source,{hal}]
----
gm.0.joint.0.home-sw-in
----

```

All the pins and parameters are updated by the following function:

----
gm.<card_no>.rs485
----

It should be added to servo thread or other thread with larger period to avoid CPU overload.
Every RS485 module pin and parameter name begins as follows:

----
gm.<card_no>.rs485.<module ID>
----

where _<module ID>_ is from 00 to 15.

.Relay output module pins
[width="80%",options="header",cols="<3,^2,<6"]
|===
| Pins                 | Type and direction | Pin description
m| .relay-<0-7>        | (bit, Out)         | Output pin for relay
|===

.Relay output module parameters
[width="80%",options="header",cols="<3,^2,<6"]
|===
| Parameters           | Type and direction | Parameter description
m| .invert-relay-<0-7> | (bit, R/W)         | Negate relay output pin
|===

.HAL example
----
#gm.0.rs485.0.relay-0  # First relay of the node.
# gm.0  # Identifies the first GM6-PCI motion control card (PCI card address = 0)
#    .rs485.0  # Selects node with address 0 on the RS485 bus

> #            .relay-0 # Selects the first relay

Reviewer: >The same for two further occurrences

.Relay output module pins
[width="80%",options="header",cols="<3,^2,<6"]
|===
| Pins                 | Type and direction | Pin description
m| .relay-<0-7>        | (bit, Out)         | Output pin for relay
|===

.Relay output module parameters
[width="80%",options="header",cols="<3,^2,<6"]
|===
| Parameters           | Type and direction | Parameter description
m| .invert-relay-<0-7> | (bit, R/W)         | Negate relay output pin
|===

> .HAL example
> |===
> m|gm.0.rs485.0.relay-0 | First relay of the node.
> m|gm.0                 | Means the first GM6-PCI motion control card (PCI card address = 0)
> m|.rs485.0             | Select node with address 0 on the RS485 bus
> m|.relay-0             | Select the first relay
> |===

Reviewer 2: ```suggestion
.First relay on RS485 node 0 on card 0
[source,{hal}]
----
gm.0.rs485.0.relay-0 # First relay of the node.
                     # gm.0 - Means the first GM6-PCI motion control card (PCI card address = 0)
                     # .rs485.0 - Select node with address 0 on the RS485 bus
                     # .relay-0 - Select the first relay
----
```

Reviewer: ```
.First relay on RS485 node 0 on card 0
====
[source,{hal}]
----
gm.0.rs485.0.relay-0 
----
`gm.0` Means the first GM6-PCI motion control card (PCI card address = 0)
`.rs485.0` Select node with address 0 on the RS485 bus
`.relay-0` Select the first relay
====
```
I agree, maybe the explanation doesn't have to be in the code block.

But if we have it inside the code block, we can have it like this:
```
.First relay on RS485 node 0 on card 0
====
[source,{hal}]
----
  gm.0.rs485.0.relay-0  # First relay of the node.
# gm.0                    Means the first GM6-PCI motion control card (PCI card address = 0)
#     .rs485.0            Select node with address 0 on the RS485 bus
#             .relay-0    Select the first relay
----
```

Reviewer 2: > ```
> .First relay on RS485 node 0 on card 0
> ====
> [source,{hal}]
> ----
> gm.0.rs485.0.relay-0 
> ----
> `gm.0` Means the first GM6-PCI motion control card (PCI card address = 0)
> `.rs485.0` Select node with address 0 on the RS485 bus
> `.relay-0` Select the first relay
> ====
> ```
> 
> I agree, maybe the explanation doesn't have to be in the code block.

This may well be my preferred format :+1: , just using a `#<1><2><3>`  callouts or a simple unordered list for the 3 elements.

> But if we have it inside the code block, we can have it like this:
> 
> ```
> .First relay on RS485 node 0 on card 0
> ====
> [source,{hal}]
> ----
>   gm.0.rs485.0.relay-0  # First relay of the node.
> # gm.0                    Means the first GM6-PCI motion control card (PCI card address = 0)
> #     .rs485.0            Select node with address 0 on the RS485 bus
> #             .relay-0    Select the first relay
> ----
> ```

That's a clever one but my eyes on first sight so...

Reviewer 2: Thought about it again last night and I'm afraid that the example block makes it standout even more than the sibling without it.
Kind of a pity because it is semantically sound. Could we make something with a pinch of CSS Fu ? Like removing bold face for example block titles ? And left side border ?

> gm.<nr of card>.rs485.<module ID>
> ----
> 
> where _<module ID>_ is from 00 to 15.

Reviewer 2: ```suggestion
gm.<card_no>.rs485.<module_id>
----

Where _<module_id>_ is from 00 to 15.
```

.Digital input output module pins
[width="100%",options="header",cols="<3,^2,<6"]
|===
| Pins             | Type and direction | Pin description
m| .in-_<0-7>_     | (bit, Out)         | Input pin
m| .in-not-_<0-7>_ | (bit, Out)         | Negated input pin
|===

> .HAL example
> [cols="<3,<6"]
> |===
> m|gm.0.rs485.0.in-0 |First input of the node.
> |===
> constructed from
> [cols="<3,<6"]
> |===
> m|gm.0     | Means the first GM6-PCI motion control card (PCI card address = 0)
> m|.rs485.0 | Select node with address 0 on the RS485 bus
> m|.in-0    |Select the first digital input module
> |===

Reviewer 2: ```suggestion
.First digital input module on RS485 node 0 on card 0
[source,{hal}]
----
gm.0.rs485.0.in-0 # First input of the node.
                  # gm.0 - Means the first GM6-PCI motion control card (PCI card address = 0)
                  # .rs485.0 - Select node with address 0 on the RS485 bus
                  # .in-0 - Select the first digital input module
----
```

Author: Reversed towards something reasonably similar.

gm.<nr of card>.rs485.<module ID>
----

where _<module ID>_ is from 00 to 15.

.Digital input output module pins
[width="100%",options="header",cols="<3,^2,<6"]
|===
| Pins             | Type and direction | Pin description
m| .in-_<0-7>_     | (bit, Out)         | Input pin
m| .in-not-_<0-7>_ | (bit, Out)         | Negated input pin
|===

.HAL example
|===
m|gm.0.rs485.0.in-0 |First input of the node.

> 2+|constructed from

Reviewer: I would exclude this from the table, this looks more like a formatting error  :innocent: 
Also ff.

> gm.<nr of card>.rs485.<module ID>
> ----
> 
> where _<module ID>_ is from 00 to 15.

Reviewer 2: ```suggestion
gm.<card_no>.rs485.<module_id>
----

Where _<module_id>_ is from 00 to 15.
```

.DAC & ADC module pins
[width="100%",options="header",cols="<3,^2,<6"]
|===
| Pins                 | Type and direction | Pin description
m| .adc-_<0-7>_        | (float, Out)       | Value of ADC input in Volts.
m| .dac-enable-_<0-3>_ | (bit, In)          |
Enable DAC output. When enable is false then DAC output is set to 0.0&thinsp;V.
m| .dac-_<0-3>_        | (float, In)        | Value of DAC output in Volts.
|===

.DAC & ADC module parameters
[width="100%",options="header",cols="<3,^2,<6"]
|===
| Parameters               | Type and direction | Parameter description
m| .adc-scale-_<0-7>_      | (float, R/W)       |
The input voltage will be multiplied bymscale before being output to .adc- pin.
m| .adc-offset-_<0-7>_     | (float, R/W)       |
Offset is subtracted from the hardware input voltage after the scale multiplier has been applied.
m| .dac-offset-_<0-3>_     | (float, R/W)       | Offset is added to the value before the hardware is updated.
m| .dac-high-limit-_<0-3>_ | (float, R/W)       | Maximum output voltage of the hardware in Volts.
m| .dac-low-limit-_<0-3>_  | (float, R/W)       | Minimum output voltage of the hardware in Volts.
|===

> .HAL example
> [cols="<3,<6"]
> |===
> m|gm.0.rs485.0.adc-0 | First analogue channel of the node.
> |===
> constructed from
> [cols="<3,<6"]
> |===
> m|gm.0     | Means the first GM6-PCI motion control card (PCI card address = 0)
> m|.rs485.0 | Select node with address 0 on the RS485 bus
> m|.adc-0   | Select the first analogue input of the module
> |===

Reviewer 2: ```suggestion
.First analog channel on RS485 node 0 on card 0
[source,{hal}]
----
gm.0.rs485.0.adc-0 # First analogue channel of the node.
                   # gm.0 - Means the first GM6-PCI motion control card (PCI card address = 0)
                   # .rs485.0 - Select node with address 0 on the RS485 bus
                   # .adc-0 - Select the first analogue input of the module
----
```

Author: The original was better, I think.

m| .enc-rawcounts  | (s32, Out)         | The raw count is the counts, but unaffected by reset.
m| .enc-position   | (float, Out)       | Position in scaled units (=.enc-counts/.enc-position-scale).
m| .in-_<0-7>_     | (bit, Out)         | Input pin
m| .in-not-_<0-7>_ | (bit, Out)         | Negated input pin
|===

.Teach Pendant module parameters
[options="header",cols="<3,^2,<6"]
|===
| Parameters            | Type and direction | Parameter description
m| .adc-scale-_<0-5>_   | (float, R/W)       | The input voltage will be multiplied by scale before being output to .adc- pin.
m| .adc-offset-_<0-5>_  | (float, R/W)       | Offset is subtracted from the hardware input voltage after the scale multiplier has been applied.
m| .enc-position-scale  | (float, R/W)       | Scale in per length unit.
|===

.HAL example
[cols="<3,<6"]
|===
| gm.0.rs485.0.adc-0 | First analogue channel of the node.
|===
constructed from
[cols="<3,<6"]
|===
m| gm.0     | Means the first GM6-PCI motion control card (PCI card address = 0)
m| .rs485.0 | Select node with address 0 on the RS485 bus
m| .adc-0   | Select the first analogue input of the module

> |===

Reviewer 2: ```suggestion
.First analog channel on RS485 node 0 on card 0
[source,{hal}]
----
gm.0.rs485.0.adc-0 # First analogue channel of the node.
                   # gm.0 - Means the first GM6-PCI motion control card (PCI card address = 0)
                   # .rs485.0 - Select node with address 0 on the RS485 bus
                   # .adc-0 - Select the first analogue input of the module
----

Author: Hm. I then prefer the original.

Reviewer: What do you think of this: https://github.com/LinuxCNC/linuxcnc/pull/2037#discussion_r980201898?

Reviewer 2: > Hm. I then prefer the original.

Not having it indented makes it unobvious that these lines pertains to the first one because they don't differ much from another commented line.

Author: > #2037 (comment)

The indented one is very nice, indeed. A conditione sine qua non for me is that code-blocks can be copied in toto and are executable. So, the non-executable parts need to be commented "away". You seem to have addressed that, so I am fine with it.

Author: I think to have addressed this now.
